docs/reference/api/platform/Error.md

```diff
@@ -1,14 +1,123 @@
 
-Mbed OS provides an error function to output messages to `STDIO` at runtime when the system encounters a fatal error and the Application calling the error function will be terminated. Note that the error function outputs the error message in [debug and develop builds](/docs/development/tools/build-profiles.html) only. In release builds, the error function does not generate any `STDIO` output, but the application is still terminated. `mbed_error.h` declares the error function.
+Conceptually, Error Handling is a platform service implemented by platform layer of Mbed OS and provides the following:
+1. Provides System Defined Error Codes and Status values in `mbed_error.h`
+1. Provides APIs for Error Construction, Reporting and Error history(see below sections), callable from anywhere in the system.
+1. Provides ISR-safe and Thread-safe error handling APIs.
+1. Provides mechanisms for extending the error status definitions.
 
-### Error function reference
+### Error Status Usage
+Mbed OS pre-defines a set of system errors and **Error Status** is defined as a 32-bit Signed Integer in **NEGATIVE space**. The error status values are represented by **mbed_error_status_t** type.
 
+Error Status can be used  
+- To indicate return status from function calls
+  - Returning 0(=**MBED_SUCCESS**) or any positive number would be considered success.
+  - A predefined status code named **MBED_SUCCESS**(=0) is available to indicate success.
+
+- As system error codes
+  - Used to report error conditions to the OS error handling sub-system.
+
+### Error Status Types and Error Code Ranges
+Mbed OS error status encoding defines three types of error status values. They are as below. Note that type of error status is encoded into the 32-bit signed integer.
+- System Errors
+- Posix Errors
+- Custom Defined Errors
+
+And for each error status type, we have defined a range of error codes which forms the least significant 16-bits of 32-bit error status values. Based on error code definitions below are the error code ranges for each error type.
+- Posix Error Codes: 1 to 255 (see **Posix Error Codes Support** section below)
+- System Error Codes: 256 to 4095
+- Custom Error Codes: 4096 to 65535 
+
+Its important to understand the difference between the terminologies "**Error Code**" and "**Error Status**". "**Error Status**" is the 32-bit signed integer in **NEGATIVE space** used by applications to report the error, and "**Error Code**" is the least significant 16-bits of 32-bit Error Status values indicating the error situation(For example, Out of memory). All pre-defined error status definitions start with **MBED_ERROR_** prefix and all pre-defined error code definitions start with **MBED_ERROR_CODE_**.
+Error codes are not used by themselves, they are always encoded into error status values when reporting or representing errors in the system.
+
+Accordingly, the Error Status Ranges are as below for each type:
```

Comment:
Would a small image showing the bit fields be helpful to have here (a la http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0211h/ch02s08s01.html)